sync/limiter.go

```diff
@@ -0,0 +1,87 @@
+package sync
+
+import (
+	"sync"
+)
+
+// A Limiter restricts access to a resource.
+type Limiter struct {
+	limit    int
+	current  int
+	requests chan struct{}
+	cond     *sync.Cond
+}
+
+// Request blocks until n units are available. If n is greater than m's limit,
+// Request blocks until all of m's units have been released.
+//
+// Request is unbiased with respect to n: calls with small n do not starve
+// calls with large n.
+//
+// Request returns true if the request was canceled, and false otherwise.
+func (l *Limiter) Request(n int, cancel <-chan struct{}) bool {
+	// wait until our request is "first in line"
+	select {
+	case <-cancel:
+		return true
+	case token := <-l.requests:
+		// return token when we're done
+		defer func() { l.requests <- token }()
+	}
```

Comment:
this is basically just a mutex; the reason I implemented it with a channel is so that we can `select` to provide cancellation.
I also explored using the `sync.Cond` for this purpose (setting/checking a `locked` boolean var), but found it to be unwieldy.

Really hoping we get lock-channels in an upcoming Go release...